hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/ProvidesHoodieConfig.scala

```diff
@@ -89,6 +89,62 @@ trait ProvidesHoodieConfig extends Logging {
       defaultOpts = defaultOpts, overridingOpts = overridingOpts)
   }
 
+  /**
+   * Get the insert operation.
+   * See if we are able to set bulk insert, else use deduceOperation
+   */
+  private def getOperation(isPartitionedTable: Boolean,
+                           isOverwritePartition: Boolean,
+                           isOverwriteTable: Boolean,
+                           insertModeSet: Boolean,
+                           dropDuplicate: Option[String],
+                           enableBulkInsert: Option[String],
+                           isInsertInto: Boolean,
+                           isNonStrictMode: Boolean,
+                           hasPrecombineColumn: Boolean): String = {
+    val notSetToNonStrict = !insertModeSet || isNonStrictMode
+    //if options are not set, we assume they are configs to do bulk insert
+    (isInsertInto, notSetToNonStrict, enableBulkInsert.getOrElse("true"),
+      dropDuplicate.getOrElse("false"), isOverwritePartition, isPartitionedTable) match {
+      case (true, true, "true", "false", false, _) => BULK_INSERT_OPERATION_OPT_VAL
+      case (true, true, "true", "false", true, false) => BULK_INSERT_OPERATION_OPT_VAL
+
+      //if config is set such that we cant make it bulk insert, we need to use defaults for unset configs
+      case _ => deduceOperation(enableBulkInsert.getOrElse(SQL_ENABLE_BULK_INSERT.defaultValue).toBoolean,
+        isOverwritePartition, isOverwriteTable, dropDuplicate.getOrElse(INSERT_DROP_DUPS.defaultValue).toBoolean,
+        isNonStrictMode, isPartitionedTable, hasPrecombineColumn)
+    }
+  }
+
+  /**
+   * Deduce the insert operation
+   */
+  private def deduceOperation(enableBulkInsert: Boolean, isOverwritePartition: Boolean, isOverwriteTable: Boolean,
+                              dropDuplicate: Boolean, isNonStrictMode: Boolean, isPartitionedTable: Boolean,
+                              hasPrecombineColumn: Boolean): String = {
+    (enableBulkInsert, isOverwritePartition, isOverwriteTable, dropDuplicate, isNonStrictMode, isPartitionedTable) match {
```

Comment:
Is there any code changes here or we just moved the code around?

Reply:
No changes